common/src/notifications.rs

```diff
@@ -41,3 +116,142 @@ pub fn set_badge(count: u32) -> Result<(), String> {
+    #[cfg(target_os = "windows")]
+    {
+        // Notify-rust does not support windows actions so we use the underlying system directly
+        // See https://gist.github.com/allenbenz/a0fb225aef43df4b1be1c005fb4c2811 for general idea
+        let duration = match notification.timeout {
+            notify_rust::Timeout::Default => "duration=\"short\"",
+            notify_rust::Timeout::Never => "duration=\"long\"",
+            notify_rust::Timeout::Milliseconds(t) => {
+                if t >= 25000 {
+                    "duration=\"long\""
+                } else {
+                    "duration=\"short\""
+                }
+            }
+        };
+        //TODO set proper app id
+        let app_id = POWERSHELL_APP_ID.to_string();
+        let template_binding = "ToastGeneric";
+        let actions = format!(
+            r#"<action content="{}" arguments="{}"/>"#,
+            &get_local_text(&action_id),
+            &action_id
+        );
+
+        let toast_xml = windows::Data::Xml::Dom::XmlDocument::new().unwrap();
+        toast_xml
+            .LoadXml(&windows::runtime::HSTRING::from(format!(
+                "<toast {} {}>
+                    <visual>
+                        <binding template=\"{}\">
+                        {}
+                        {}{}{}
+                        </binding>
+                    </visual>
+                    {}
+                    <actions>
+                        {}
+                    </actions>
+                </toast>",
+                duration,
+                String::new(), //Scenario
+                template_binding,
+                &notification.icon,
+                &notification.summary,
+                notification.subtitle.as_ref().map_or("", AsRef::as_ref),
+                &notification.body,
+                r#"<audio silent='true'/>"#, //Already handled in uplink
+                actions
+            )))
+            .unwrap();
+
+        // Create the toast
+        let toast_notification =
+            windows::UI::Notifications::ToastNotification::CreateToastNotification(&toast_xml)
+                .unwrap();
+
+        toast_notification
+            .Activated(windows::Foundation::TypedEventHandler::new(
+                move |_sender, result: &Option<windows::runtime::IInspectable>| {
+                    let event: Option<
+                        windows::runtime::Result<
+                            windows::UI::Notifications::ToastActivatedEventArgs,
+                        >,
+                    > = result.as_ref().map(windows::runtime::Interface::cast);
+                    let arguments = event
+                        .and_then(|val| val.ok())
+                        .and_then(|args| args.Arguments().ok());
+                    if let Some(val) = arguments {
+                        if val.to_string_lossy().eq(&action_id) {
+                            log::trace!("windows action activated {:?}", val);
```

Comment:
the message being passed to `trace!` seems inaccurate. the thing being "activated" is a toast notification, not a "windows activation"